src/main/scala/is/hail/expr/AST.scala

```diff
@@ -404,7 +404,7 @@ case class ArrayConstructor(posn: Position, elements: Array[AST]) extends AST(po
       CompilationHelp.arrayToWrappedArray(convertedArray)
 
   def toIR(agg: Option[String] = None): Option[IR] = for {
-    irElements <- anyFailAllFail(elements.map(_.toIR(agg)))
+    irElements <- anyFailAllFail(elements.map(_.toIR(agg))).filter(_ => elements.map(_.`type`).distinct.length == 1)
```

Comment:
Shouldn't this get checked on the python side?  When the AST goes away who's going to catch this error?

Reply:
this is a problem with the IR. It doesn't support heterogeneous array declarations right now, even though the AST (and Python interface) do.

Reply:
heterogeneous meaning `[0, 1.0]` for example.

Reply:
Ah.  The IR shouldn't: someone else should put in type conversions.  But this is a fine way to punt.